File: lsf-configurator/frontend/src/models/dto.ts
export interface FunctionAppCreateDto {
  name: string;
  runtime: string;
  components: {
    name: string;
    memory: number;
    runtime: number;
    files: string[];
  }[];
  links: {
    from: string;
    to: string;
    invocationRate: {
      min: number;
      max: number;
    };
    dataDelay: number;
  }[];
  sourcePath?: string;
  latencyLimit: number;
  platformManaged: boolean;
}

export interface FunctionCompositionCreateDto {
  functionAppId: string;
  components: string[];
}

export interface DeploymentCreateDto {
  functionCompositionId: string;
  node: string;
  namespace: string;
  routingTable: Record<string, Array<{ to: string; function: string }>>;
}

export interface BulkCreateRequest {
  functionApp: FunctionAppCreateDto;
  functionCompositions: FunctionCompositionBulkCreateDto[];
  deployments: DeploymentBulkCreateDto[];
}

export interface FunctionCompositionBulkCreateDto {
  id: string; // here ID is a temporary ID, given by the user for finding matching objects in bulk requests
  components: string[];
  files: string[];
}

export interface DeploymentBulkCreateDto {
  id: string; // here ID is a temporary ID, given by the user for finding matching objects in bulk requests
  functionCompositionId: string;
  node: string;
  namespace: string;
  routingTable: Record<string, Array<{ to: string; function: string }>>;
}

File: lsf-configurator/api/dto.go
package api

import "lsf-configurator/pkg/core"

type FunctionAppCreateDto struct {
	Name            string               `json:"name"`
	Runtime         string               `json:"runtime"`
	Components      []core.Component     `json:"components"`
	Links           []core.ComponentLink `json:"links"`
	LatencyLimit    int                  `json:"latency_limit"`
	PlatformManaged bool                 `json:"platform_managed"`
}

type FunctionCompositionCreateDto struct {
	FunctionAppId string   `json:"function_app_id"`
	Components    []string `json:"components"`
	Image         string   `json:"image"`
}

type DeploymentCreateDto struct {
	FunctionCompositionId string            `json:"function_composition_id"`
	Node                  string            `json:"node"`
	Namespace             string            `json:"namespace"`
	RoutingTable          core.RoutingTable `json:"routing_table"`
}

type BulkCreateRequest struct {
	FunctionApp          FunctionAppCreateDto               `json:"function_app"`
	FunctionCompositions []FunctionCompositionBulkCreateDto `json:"function_compositions"`
	Deployments          []DeploymentBulkCreateDto          `json:"deployments"`
}

type FunctionCompositionBulkCreateDto struct {
	TempId     string   `json:"id"`
	Components []string `json:"components"`
	Files      []string `json:"files"`
	Image      string   `json:"image"`
}

type DeploymentBulkCreateDto struct {
	TempId                    string            `json:"id"`
	TempFunctionCompositionId string            `json:"function_composition_id"`
	Node                      string            `json:"node"`
	Namespace                 string            `json:"namespace"`
	RoutingTable              core.RoutingTable `json:"routing_table"`
}
